Making Fleetspeak server metrics scrapable by Prometheus

tsehori opened the pull request:
Added a new Collector struct, PrometheusStatsCollector, that integrates with Prometheus. Modified `config.proto` to have flags related to stats, and added corresponding code in `components.go` that parses it and starts a stats server accordingly. Also added a version of `config.pb.go` recompiled with protoc-gen-go that goes with the modified `config.proto`.

This PR also solves #191.

bgalehouse commented on fleetspeak/src/server/stats.go:
```diff
@@ -59,6 +62,132 @@ func (s noopStatsCollector) ResourceUsageDataReceived(cd *db.ClientData, rud mpb
+// A prometheusStatsCollector is an implementation of a Collector interface.
+// It exports stats to a Prometheus HTTP handler, which are exposed at :2112/metrics
+// and are scrapable by Prometheus.
+type prometheusStatsCollector struct{}
```
There is a lot of additional metrics that can be added here, especially as histograms. Obvious things: database operation time, message processing time, client memory usage, client cpu usage, and probably a lot more that I'm forgetting.

Also, most of these things can and should be broken down by one or more parameter. If you are running multiple services it is very helpful to know which service's message (or which type of which service's message) is erroring, not that messages are erroring. It is always helpful to know which database operations are slow, etc, etc.

Glancing at the prometheous godoc, I think you want the *Vec types for most of these.

Anyway, I think this is probably fine to pull as a starting point, but I don't think it is complete enough to close #191

tsehori replied:
Regarding breaking down the more parameters - I am not sure how those would be gathered. For instance, in the `MessageSaved` [metric](https://github.com/google/fleetspeak/blob/7ccefefe98e1e92af7559fb1d849bb1eb2ea3c48/fleetspeak/src/server/stats.go#L153), a lot of data is gathered, like the service used, the message type and so on. These can vary depending on the system being used (which is not necessarily GRR), and so unlike the rather simple [Counters](https://github.com/google/fleetspeak/blob/7ccefefe98e1e92af7559fb1d849bb1eb2ea3c48/fleetspeak/src/server/stats.go#L83) I have implemented for it so far, these kind of metrics would have to somehow dynamically create new Counters/Gauges/etc.
Am I going about this the right way, or should it be simpler than that?

I'll make sure to add metrics in a different, dedicated PR so it's easier to review and maintain. I'll also link to this PR and discussion for reference.

mbushkov replied:
My understanding is that Prometheus labels is what has to be used here (see 
https://prometheus.io/docs/concepts/data_model/#metric-names-and-labels and https://prometheus.io/docs/practices/naming/).

Labels effectively introduce additional dimensions to the metric. Therefore it's important that each label has a finite number of possible values. This assumption holds for thinks like service name (there's a finite number of services in a particular fleetspeak deployment), database operation type, etc.

For example, in case of [MessageSaved](https://github.com/google/fleetspeak/blob/298e9ae7573cf0c2f29b0c1e68a0a11370259c53/fleetspeak/src/server/stats/collector.go#L94), `service`, `messageType` and `forClient` should likely be labels. While `savedPayloadBytes` is what will get accumulated in the metric value.

In any case, figuring this out in a separate PT sounds good.

bgalehouse replied:
I've never used Prometheus, but it is similar to what we use internally, and internally, we find it  critical to split metrics by various attributes.

Again, looking through the docs, the *Vec Prometheus types seem to be what we need here, look at the example code in: https://godoc.org/github.com/prometheus/client_golang/prometheus#CounterVec.With

mbushkov replied:
+1 to Ben's suggestion. Please see the example at https://godoc.org/github.com/prometheus/client_golang/prometheus#CounterVec:

```
httpReqs := prometheus.NewCounterVec(
    prometheus.CounterOpts{
        Name: "http_requests_total",
        Help: "How many HTTP requests processed, partitioned by status code and HTTP method.",
    },
    []string{"code", "method"},
)
prometheus.MustRegister(httpReqs)

httpReqs.WithLabelValues("404", "POST").Add(42)
```

Here `code` and `method` are Prometheus labels.

mbushkov commented on fleetspeak/src/server/stats.go:
```diff
@@ -59,6 +62,132 @@ func (s noopStatsCollector) ResourceUsageDataReceived(cd *db.ClientData, rud mpb
+var (
+	// Metric collectors for PrometheusStatsCollector struct
+	messagesIngested = promauto.NewCounter(prometheus.CounterOpts{
+		Name: "fleetspeak_messages_ingested_total",
```
+1 to Ben's opinion in his [review](https://github.com/google/fleetspeak/pull/247#discussion_r434807335)

Fleetspeak has a notion of "service". I.e. GRR is a service in fleetspeak terms. A single Fleetspeak deployment may act as a communication middleware for a number of services. Every ingested message has a source/destination attributes, indicating the service responsible for the message (https://github.com/google/fleetspeak/blob/master/fleetspeak/src/common/proto/fleetspeak/common.proto#L9). (@bgalehouse, btw - do I understand it correctly that the service in source/destination is always the same?) We should include the service into the metric, so that later in Grafana we can filter by it.

tsehori replied:
See [my reply](https://github.com/google/fleetspeak/pull/247#discussion_r435077477) to Ben's review.
Just to clarify - when you say that a single Fleetspeak deployment may act as a communication middleware for a number of services, do you mean concurrently? As in, multiple services (GRR server/s and multiple others) could utilize the same Fleetspeak deployment?

mbushkov replied:
Yes, multiple services can utilize the same fleetspeak deployment at the same time. Each service is identified in fleetspeak configuration by its name and fleetspeak uses the configuration to route the messages to proper client/server plugins.

bgalehouse replied:
One is the service name on the client, one is the service name on the server. Which is source and which is dest depends in the obvious way on which direction the message is going.

In the case of GRR, both the client and server services are called "GRR". However, FS does not require that they be the same, and in other applications we've found making them the same to be confusing. The natural name for the daemon on the client might not be the same as the natural name for the message's next stop in the cloud.

So we should not assume that they are the same in FS. For messages received from the server from a client, the destination service name is a server service name, and this is a good parameter to break down statistics about messages arriving from clients for the server.

mbushkov commented on fleetspeak/src/server/stats.go:
```diff
@@ -59,6 +62,132 @@ func (s noopStatsCollector) ResourceUsageDataReceived(cd *db.ClientData, rud mpb
+		Help: "The total number of times a client polls the Fleetspeak server.",
+	})
+
+	datastoreOperationsCompleted = promauto.NewCounter(prometheus.CounterOpts{
```
+1 to Ben's comment. One global counter for completed datastore operations is not very useful. We should be able to filter/group by different operations in Grafana.

tsehori commented on fleetspeak/src/server/components/proto/fleetspeak_components/config.proto:
```diff
@@ -75,3 +78,13 @@ message AdminConfig {
+message StatsConfig {
+  // Stats collector to be used. Optional, but recommended.
+  // Defaults to no stats collector.
+  string collector = 1;
```
Yes, that was one thing I was not entirely sure about. I figured that if a user wishes to use a different type of collector, which is not Prometheus, it would be easier to later integrate it without having to add an additional flag (such as `collector`) to the config file.
But you're right, and I also think that for the open source version it suffices to have just Prometheus.

bgalehouse replied:
The original design intent is that advanced users can define and compile in a collector to bridge to any monitoring system that they want to use. Non-advanced users can pick a standard implementation without recompilation, using this components proto.

I think that having a string here is strange - enum would be better - also that it is probably fine to no worry about system selection here. When we have a another case, we can decide how to adjust the config proto. e.g. if it also only requires a port, we could add an enum selector that defaults to prometheus. If it requires something else, we need to rework the proto more.

tsehori commented on fleetspeak/src/server/components/proto/fleetspeak_components/config.proto:
```diff
@@ -75,3 +78,13 @@ message AdminConfig {
+  string collector = 1;
+
+  // The bind address to listen on for connections, e.g. ":443" or
+  // "localhost:1234". Optional, defaults to ":2112".
```
I agree that mentioning port 443 might be misleading and/or confusing, so I'll make sure to stick to an obviously arbitrary port like 2112 and get rid of the notion of a "default" port to fit the logic.
The reason for using 2112 as default was merely inspired by the Prometheus docs to [instrumenting a Go application](https://prometheus.io/docs/guides/go-application/#how-go-exposition-works) (such as Fleetspeak), where the example port was 2112 as well.

bgalehouse replied:
Mentioning it as a default probably makes sense. But the port shouldn't be publicallly accessible, so it is probably best if it requires some intent on the part of the admin setting things up to open it. Especially if it is bound to a public address ":2112" instead of "localhost:2112", we should make it hard to miss the need to firewall it.

tsehori replied:
> Mentioning it as a default probably makes sense. But the port shouldn't be publicallly accessible, so it is probably best if it requires some intent on the part of the admin setting things up to open it. Especially if it is bound to a public address ":2112" instead of "localhost:2112", we should make it hard to miss the need to firewall it.

I'm not sure what you mean - do you think that in the comments, an example should be in the form of "localhost:2112" rather than plain ":2112"?

bgalehouse replied:
http.ListenAndServer(":2112") listens on all interfaces. http.ListenAndServer("localhost:2112") listens on the loopback interface. The decision has security implications - the first is safer, and great for demos, but obviously isn't what you want in a large install.

In some installations the host might have both a public and private interface, with the private interface/subnet used to communicate purely with other servers. In this case, they might want to bind to "<private address>:2112".

This is why I try to always use bind address as the controlling parameter, not port number. This standard is suggested by the go libraries, and it is a good one.

So yes, the method you linked should change and the protocol buffer field should be renamed accordingly.

tsehori commented:
I committed the part of the changes required in 0e7e4a9, regarding all comments about the configuration files (`config.proto` and the corresponding compiled `config.pb.go`) and `components.go`. I removed the string flag of choosing Prometheus over a different stats collector (as discussed, there is no need for one at this time). Per @bgalehouse's [comment](https://github.com/google/fleetspeak/pull/247#discussion_r435334343), I modified it so only a full bind address can be used, and not a port alone.
As @mbushkov [has mentioned](https://github.com/google/fleetspeak/pull/247#pullrequestreview-423916141), I think it is best that this commit (and effectively, this PR) will only cover the configuration aspect of Prometheus. I will leave the basic metrics implemented in this PR as is for merging as a start, and later submit a PR dedicated to expanding those metrics as discussed in @bgalehouse's [review](https://github.com/google/fleetspeak/pull/247#pullrequestreview-423870217) and @mbushkov's [comment](https://github.com/google/fleetspeak/pull/247#discussion_r434855822).

bgalehouse commented on fleetspeak/src/server/components/proto/fleetspeak_components/config.proto:
```diff
@@ -80,11 +80,9 @@ message AdminConfig {
-  // The bind address to listen on for connections, e.g. ":443" or
-  // "localhost:1234". Optional, defaults to ":2112".
-  string port = 2;
+  // The bind address to listen on for connections in the
```
nit: suggest mentioning the type of connection, e.g. "to listen on for Prometheus http metric collection"

tsehori replied:
Fixed in 1d8be41.

bgalehouse commented on fleetspeak/src/server/components/components.go:
```diff
@@ -139,14 +139,11 @@ func MakeComponents(cfg cpb.Config) (*server.Components, error) {
 			statsCollector = server.PrometheusStatsCollector{}
 			http.Handle("/metrics", promhttp.Handler())
-			go http.ListenAndServe(":" + portToExportStats, nil)
+			go http.ListenAndServe(addressToExportStats, nil)
```
I just realize that we need to figure out if this use of the http global ServerMux interacts with the http server that is run over in the notification code: https://github.com/google/fleetspeak/blob/master/fleetspeak/src/server/components/notifications/http.go

Skimming to docs, I think both handlers must end up on the same DefaultServeMux, and then I'm really not sure what running Serve twice at the same time does.

The two ways to clean this up are 1) create actual http.Server objects in both cases, so that they end up with separate multiplexers and start/stop independently, or 2) organize a common http server for internal coordination and metrics (possibly based on the default mux). configure its bind address in one place and attach both handlers to it.

tsehori replied:
> I just realize that we need to figure out if this use of the http global ServerMux interacts with the http server that is run over in the notification code: https://github.com/google/fleetspeak/blob/master/fleetspeak/src/server/components/notifications/http.go
> 
> Skimming to docs, I think both handlers must end up on the same DefaultServeMux, and then I'm really not sure what running Serve twice at the same time does.
> 
> The two ways to clean this up are 1) create actual http.Server objects in both cases, so that they end up with separate multiplexers and start/stop independently, or 2) organize a common http server for internal coordination and metrics (possibly based on the default mux). configure its bind address in one place and attach both handlers to it.

Me and @mbushkov have taken a look into [this article](https://echorand.me/posts/golang-dissecting-listen-and-serve/#using-your-own-handler-with-listenandserve) which suggested to create a separate mux other than DefaultServeMux and attach the Prometheus HTTP handler to it.
Please have a look at 4c85c00.

bgalehouse replied:
Yeah, that fixes the problem locally, and is good enough to get started with. Long term we should think about whether we want one or two ports for these internal communications purposes. (there is also a more involved example alread in the code at https://github.com/google/fleetspeak/blob/master/fleetspeak/src/server/https/https.go#L95)

mbushkov commented:
Changes look good to me. @bgalehouse , would you mind taking a look so that we can merge this?

bgalehouse commented on fleetspeak/src/server/components/components.go:
```diff
@@ -140,6 +157,7 @@ func MakeComponents(cfg cpb.Config) (*server.Components, error) {
 		},
 		Communicators: []comms.Communicator{comm},
 		Authorizer:    auth,
+		Stats:		   statsCollector,
```
nit:alignment
(we occasionally run everything through "gofmt -s")

bgalehouse commented:
I just realized that this PR #247 make the core Fleetspeak library (through stats.go) dependent on Prometheus. This is not what we want. We want it to be possible to build a FS server that does not depend on the prometheus.

Please create a subpackage of components to contain your new StatsCollector and ensure that all use of the prometheus client libraries goes through it.

tsehori commented:
Do you also refer to the use of `promhttp` in components.go?

bgalehouse commented:
The idea is that each subdirectory of component is an (optional) plugin which a FS installation might or might not want to use. components.go exists to bring these together and allow dynamic selection of these components. So there are two way so to use component:

1) Use the provided distribution binary, or otherwise import components.go. In this case all these optional plugins will be linked in.
2) Select which components are used by compiling a main which instantiates a server object with your own static component choices. This allows the installation to select exactly what is needed, and only link in those plugins (and their dependencies) that are needed. This also makes it possible to define 'your' own plugin components - components specific to your internal technology.

So, a subdirectory of components/, perhaps components/prometheus, should define a prometheus base stats collector. The other subdirectories of components/ should not depend on or be dependent on this directory. The collectors.go file and proto can and should continue to depend on it, and make it possible to dynamically choose to install such a collector when starting a FS server - promhttp is probably fine. But somebody standing up a FS server directly (option 2) should not need to depend on the prometheus client libraries, unless they explicitly pull in the components/prometheus package.

tsehori commented:
Thanks for the explanation @bgalehouse! I'll make sure to handle it in a separate PR, after #258 is merged (so there won't be any significant merge conflicts).
